app/Jobs/CreateQueryserviceBatchesJob.php

```diff
@@ -0,0 +1,86 @@
+<?php
+
+namespace App\Jobs;
+
+use App\EventPageUpdate;
+use App\QsBatch;
+use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Config;
+
+class CreateQueryserviceBatchesJob extends Job
+{
+    private const NAMESPACE_ITEM = 120;
+    private const NAMESPACE_PROPERTY = 122;
+    private const NAMESPACE_LEXEME = 146;
+
+    private int $entityLimit;
+
+    public function __construct()
+    {
+        $this->entityLimit = Config::get('wbstack.qs_batch_entity_limit');
+    }
+
+    public function handle(): void
+    {
+        DB::transaction(function () {
+            // Get ID of the latest batch that we have created (or 0)
+            $maxIdBatch = QsBatch::orderBy('id', 'desc')->first();
+            if ($maxIdBatch) {
+                $batchesUpToEventId = $maxIdBatch->eventTo;
+            } else {
+                // If there are no batches then things just hav not really started yet
+                $batchesUpToEventId = 0;
+            }
+
+            // Get events after the point that batch was created
+            // TODO maybe filter by NS here?
+            $events = EventPageUpdate::where('id', '>', $batchesUpToEventId)->get();
+
+            $wikiBatchesEntities = [];
+            $lastEventId = 0;
+            foreach ($events as $event) {
+                if (
+                    $event->namespace == self::NAMESPACE_ITEM ||
+                    $event->namespace == self::NAMESPACE_PROPERTY ||
+                    $event->namespace == self::NAMESPACE_LEXEME
+                ) {
+                    $wikiBatchesEntities[$event->wiki_id][] = $event->title;
+                }
+                if ($event->id > $lastEventId) {
+                    $lastEventId = $event->id;
+                }
+            }
+
+            $notDoneBatches = QsBatch::where([
+                ['done', '=', 0], ['pending_since', '=', null], ['failed', '=', false]
+            ])->get();
+
+            // Insert the newly created batches into the table...
+            foreach ($wikiBatchesEntities as $wikiId => $entityBatch) {
+                // If we already have a not done batch for this same wiki, then merge that into a new batch
+                foreach ($notDoneBatches as $qsBatch) {
+                    if ($qsBatch->wiki_id == $wikiId) {
```

Comment:
I'm struggling a bit to parse the logic here; might just be me being slow though.

Does this code take into account the situation where there are many Events unbatched (and that length is already longer than the batch limit?) To me this only appears to prevent us merging batches and making one that is too big. If so maybe we want to account for that?

Reply:
Correct, this would not do anything on a batch of size 100 that already exists in the database. It would just not append further items to it. The batch of size 100 would remain in the db until processed though.

I wasn't too sure if this is in scope here, so I left it out. If we want this, I'd rather put it in a manual one-off command that does break down existing batches that have grown too big into properly sized ones.

Reply:
Added such a command in https://github.com/wbstack/api/pull/685/commits/a58d8a03b7075a4e232900e97ffa251d80bd3181

Reply:
About "unbatched" entries: if we trust the code to do what we want it to do, this should never happen and so I'm not sure if it makes sense to add code that would cater for a case that other code shouldn't even create (not saying that it could happen though, that's correct).

Reply:
I think I'm mostly worried about "unbatched" entries as you're calling them here. If I understand rightly (I may not) then this can happen in a few cases:

- crazy high edit rate where someone makes more than batch limit entries in under 1 min
- contention on the job runner - this might be scheduled every minute but if it doesn't actually get run due to something else in the way then this could happen?
- something happens such that *this* exact transaction fails to commit (basically what we saw yesterday) when the DB is giving some error

Am I being paranoid here? I think I could be but this exact situation caught me out in trying to fix up the site yesterday

I wouldn't block this patch on this since this is clearly better and the "hard" upper limit is clearly orders of magnitude bigger than the batchsize we are envisioning. I just forsee us possibly getting in a state where there is a backlog after an outage of some kind and we get hit by this again.

Maybe we need to add an alert on "eventTo" vs newest event id being crazy far apart?

Reply:
I have to admit I don't really see this happening all too often, and in case it does, we end up with multiple smaller batches than it's ideal, but this is not going to break anything. It might affect speed, but it's not going to have an effect as dramatic as what we've seen.

Or to put it that way: I'd be more worried about introducing subtle bugs, especially around accidentally changing ordering, in such a reconciliation logic than I am worried about having a non-ideal distribution of small batches.

Or am I missing some scenario in which for example a wiki with 4 batches of 3 entities each being created could hurt us?

Reply:
If we wanted to do this, I'd rearchitect the entire setup sth like this:

- there is one `qs_batches` table that works as previously (i.e. no limit for batch sizes) with a fixed ordering mechanism
- calling the QsController's GET method plucks the next 10 entities off the oldest item in that `qs_batches` table and turns it into a `qs_job`
- the items in `qs_job` are then subject to the pending/retry/failed logic we already have

Then again, I'm not sure it'd be worth it.